### PR TITLE
Deprecated fastpath parameter in Categorial constructor

### LINKED ISSUE #17562: API: Deprecate Categorical fastpath argument
After https://github.com/pandas-dev/pandas/pull/16015 the `fastpath` argument to the Categorical constructor is unnecessary since you can skip the categorical validation by passing in a `CategoricalDtype`.l

### PR DESCRIPTION
**Pull request by kaichogami**
Hello everyone! 
I am new to Pandas developer community and your feedback would help me a lot. Thank you!

- [ ] closes #17562

### CODE REVIEW

File: pandas/core/categorical.py [hunk 1]
```diff
@@ -563,7 +566,7 @@ def _from_inferred_categories(cls, inferred_categories, inferred_codes,
             dtype = CategoricalDtype(cats, ordered=False)
             codes = inferred_codes
 
-        return cls(codes, dtype=dtype, fastpath=True)
+        return cls(codes, dtype=dtype)
```
Comment (kaichogami) [inline pandas/core/categorical.py]: I am quite confused on what should be done for the fastpath parameters used in class method's to create a new `Categorial`. Should I replace them with `dtype` or let the `__init__` handle `CategorialDtype` construction?

**Comment by kaichogami** [comment]
@TomAugspurger

**Comment by TomAugspurger** [comment]
Thanks @kaichogami!

I didn't quite understand what fastpath was doing earlier. It seems to only be valid when `values` is actually `codes`, so I think we should focus the effort on fixing `from_codes` (but the warning you put in place looks good and should stay).

Ok, so, we'll want to modify `Categorical.from_codes` to also take a `dtype` argument.

```python
def from_codes(cls, codes, categories, ordered=None, dtype=None):
   ...
```

We'll do the validation of `categories`, `ordered`, and `dtype`, basically refactor the first section of `__init__` to something like

```
def validate_dtype(values=None, categories=None, ordered=None, dtype=None):

        # Ways of specifying the dtype (prioritized ordered)
        # 1. dtype is a CategoricalDtype
        #    a.) with known categories, use dtype.categories
        #    b.) else with Categorical values, use values.dtype
        #    c.) else, infer from values
        #    d.) specifying dtype=CategoricalDtype and categories is an error
        # 2. dtype is a string 'category'
        #    a.) use categories, ordered
        #    b.) use values.dtype
        #    c.) infer from values
        # 3. dtype is None
        #    a.) use categories, ordered
        #    b.) use values.dtype
        #    c.) infer from values

        if dtype is not None:
            if isinstance(dtype, compat.string_types):
                if dtype == 'category':
                    dtype = CategoricalDtype(categories, ordered)
                else:
                    msg = "Unknown `dtype` {dtype}"
                    raise ValueError(msg.format(dtype=dtype))
            elif categories is not None or ordered is not None:
                raise ValueError("Cannot specify both `dtype` and `categories`"
                                 " or `ordered`.")

            categories = dtype.categories
            ordered = dtype.ordered

        elif is_categorical(values):
            dtype = values.dtype._from_categorical_dtype(values.dtype,
                                                         categories, ordered)
        else:
            dtype = CategoricalDtype(categories, ordered)
        return categories, ordered, dtype
```

And use that in both `__init__` and `from_codes`.

Then, in `from_codes`, instead of `return cls(codes, categories...)`, you'll do

```python
self = cls.__new__(cls)
self._codes = coerce_indexer_dtype(codes, dtype.categories)
self._dtype = dtype
return self
```

I *think* that'll work. Then, anywhere you see a `fastpath=True`, you should be able to change it to use `from_codes` with a dtype, and it'll be as fast as possible. Make sense?

**Comment by kaichogami** [comment]
Yes, it does. I will go change it as soon as I can. Thanks for the feedback! :smile:

**Comment by kaichogami** [comment]
Hi, I am a bit busy so if its something important somebody else can take it over, otherwise I will do it by next week. Thank you.

**Comment by TomAugspurger** [comment]
No rush on this.

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
closing as stale, if you want to work on this, pls ping.

**Pull request closed without merging**